### DISCUSSION TITLE
In what scenarios are cache headers forced to "public, max-age=0, must-revalidate" by Vercel?

### DISCUSSION THREAD

**Discussion by good-idea** [vercel/next.js#68632] [App Router]
Our project is using Next.js 14.2.5 + app router, deployed to Vercel.

We are trying to debug why pages on our site - the initial GET request to `/page` -  are all being returned with a `cache-control: public, max-age=0, must-revalidate` response header. We would expect and are hoping for it to be `max-age=600 stale-while-revalidate`.

Running a production build locally (not using vercel) gives us the results we are expecting.

The pages all have:

```
export const dynamic = 'force-static';
export const revalidate = 600; 
```

And use `async` React components to load the data. (Some of them go on to render other components which may contain their own `async` React components)

We have explicitly set `cache: 'force-cache'` on all requests made to generate the page. Some of these are GET, some are POST. These requests are given tags.

The response includes a `x-vercel-cache: HIT` heading - does this refer to the CDN cache or the Next.js data cache?

We are using middleware for localized routing (i.e. `/some-page`, `/en-US/some-page`)

We have a check for `draftMode().isEnabled` on every request - but we tried disabling this and doing so had no effect.

The documentation is not clear on when the `dynamic = 'force-static'` setting will be overridden. I understand that making a fetch request with `cache: 'no-store'`, reading cookies, or reading headers will force a page to be dynamic - but we are not doing any of this.

We see that a fresh project created by `create-next-app` and deployed to vercel is showing the same behavior.

What else could be happening that would be overriding `force-static`?

**Comment by leerob** [comment]
You're seeing SSR-like caching headers because you're turning on draft mode for every request. I was not able to reproduce with a fresh project (seeing `x-vercel-cache: HIT`, meaning it is a fully static page).

Are there any minimal reproductions?

**Reply by x0good** [reply]
1)
We don't use draft mode but also in Vercel all nextjs **static** pages return `public, max-age=0, must-revalidate`

```
access-control-allow-origin:
*
age:
384962
cache-control:
public, max-age=0, must-revalidate
content-disposition:
inline
content-encoding:
br
....
content-type:
text/html; charset=utf-8
date:
Mon, 17 Feb 2025 16:22:26 GMT
server:
Vercel
strict-transport-security:
max-age=63072000
vary:
RSC, Next-Router-State-Tree, Next-Router-Prefetch, Next-Router-Segment-Prefetch
x-matched-path:
/mission

x-nextjs-prerender:
1
x-nextjs-stale-time:
4294967294
x-vercel-cache:
HIT
x-vercel-id:
arn1::48khl-1739809346046-0719a57e9b6b
```

locally with `npm run build && npm run start` static page cache response Headers >
```
cache-control:
s-maxage=43200, stale-while-revalidate=31492800
```

2)
and second question > 
even locally when I open nextjs static page in browser: it's still waiting for server and never show in dev tools that page downloaded from browser cache. Is it expected behaviour of nextjs ?

for me it's confusing, I expect if "next build" show that page url is static -> it should be static and cached by browser automatically or at least not trigger any vercel functions during refresh page.
Or when we see `public, max-age=0, must-revalidate` + `HIT` in response header it means that NO Vercel Functions triggered and Vercel cache still works properly ?

@leerob could you explain please, I read official doc and see your videos in youtube but still not understand how test caching of static generated pages in vercel
